File: src/cavebot/vision/minimap.py
"""Minimap position detection via template matching."""

from __future__ import annotations

import cv2
import numpy as np

from cavebot.navigation.atlas import CHUNK_SIZE, Atlas


class MinimapLocator:
    def __init__(self, atlas: Atlas):
        self._atlas = atlas
        self._last_chunk: tuple[int, int, int] | None = None
        self._last_pos: tuple[int, int] | None = None

    def locate(
        self,
        minimap_snippet: np.ndarray,
        z: int,
        confidence_threshold: float = 0.7,
    ) -> tuple[int, int, float] | None:
        snippet_h, snippet_w = minimap_snippet.shape[:2]

        if np.std(minimap_snippet) == 0:
            return None

        if self._last_chunk is not None and self._last_chunk[2] == z:
            chunks = self._get_neighbor_chunks(self._last_chunk)
        else:
            chunks = self._atlas.chunk_keys_for_z(z)

        best_score = -1.0
        best_pos: tuple[int, int] | None = None
        best_chunk: tuple[int, int, int] | None = None

        for chunk_key in chunks:
            tile = self._atlas.get_color_tile(*chunk_key)
            if tile is None:
                continue
            if tile.shape[0] < snippet_h or tile.shape[1] < snippet_w:
                continue

            result = cv2.matchTemplate(tile, minimap_snippet, cv2.TM_CCOEFF_NORMED)
            _, max_val, _, max_loc = cv2.minMaxLoc(result)

            if max_val > best_score:
                best_score = max_val
                best_pos = (chunk_key[0] + max_loc[0], chunk_key[1] + max_loc[1])
                best_chunk = chunk_key

        if best_score < confidence_threshold or best_pos is None:
            if self._last_chunk is not None:
                self._last_chunk = None
                return self.locate(minimap_snippet, z, confidence_threshold)
            return None

        self._last_chunk = best_chunk
        self._last_pos = best_pos
        return (best_pos[0], best_pos[1], best_score)

    def _get_neighbor_chunks(
        self, center: tuple[int, int, int]
    ) -> list[tuple[int, int, int]]:
        cx, cy, cz = center
        neighbors = []
        for dx in [-CHUNK_SIZE, 0, CHUNK_SIZE]:
            for dy in [-CHUNK_SIZE, 0, CHUNK_SIZE]:
                key = (cx + dx, cy + dy, cz)
                if key in self._atlas.color_chunks:
                    neighbors.append(key)
        return neighbors

File: go-cavebot/internal/vision/minimap.go
package vision

import (
	"image"

	"github.com/anthropics/tibiabot/internal/navigation"
	"gocv.io/x/gocv"
)

type MinimapLocator struct {
	atlas     *navigation.Atlas
	lastChunk *navigation.ChunkKey
	// Cache converted GoCV mats to avoid repeated conversion
	matCache map[navigation.ChunkKey]gocv.Mat
}

func NewMinimapLocator(atlas *navigation.Atlas) *MinimapLocator {
	return &MinimapLocator{
		atlas:    atlas,
		matCache: make(map[navigation.ChunkKey]gocv.Mat),
	}
}

// Close releases cached GoCV Mats.
func (ml *MinimapLocator) Close() {
	for _, mat := range ml.matCache {
		mat.Close()
	}
	ml.matCache = nil
}

// Locate finds the position of the minimap snippet within the atlas.
// Returns (x, y, confidence, found).
func (ml *MinimapLocator) Locate(snippet gocv.Mat, z int, confidenceThreshold ...float64) (int, int, float64, bool) {
	threshold := 0.7
	if len(confidenceThreshold) > 0 {
		threshold = confidenceThreshold[0]
	}

	snippetH := snippet.Rows()
	snippetW := snippet.Cols()

	mean, stddev := gocv.NewMat(), gocv.NewMat()
	defer mean.Close()
	defer stddev.Close()
	gocv.MeanStdDev(snippet, &mean, &stddev)
	if stddev.GetDoubleAt(0, 0) == 0 && stddev.GetDoubleAt(1, 0) == 0 && stddev.GetDoubleAt(2, 0) == 0 {
		return 0, 0, 0, false
	}

	var chunks []navigation.ChunkKey
	if ml.lastChunk != nil && ml.lastChunk.Z == z {
		chunks = ml.getNeighborChunks(*ml.lastChunk)
	} else {
		chunks = ml.atlas.ChunkKeysForZ(z)
	}

	bestScore := float32(-1.0)
	bestX, bestY := 0, 0
	var bestChunk *navigation.ChunkKey

	mask := gocv.NewMat()
	defer mask.Close()

	for _, ck := range chunks {
		tile := ml.getTileMat(ck)
		if tile.Empty() {
			continue
		}
		if tile.Rows() < snippetH || tile.Cols() < snippetW {
			continue
		}

		result := gocv.NewMat()
		gocv.MatchTemplate(tile, snippet, &result, gocv.TmCcoeffNormed, mask)
		_, maxVal, _, maxLoc := gocv.MinMaxLoc(result)
		result.Close()

		if maxVal > bestScore {
			bestScore = maxVal
			bestX = ck.X + maxLoc.X
			bestY = ck.Y + maxLoc.Y
			c := ck
			bestChunk = &c
		}
	}

	if bestScore < float32(threshold) {
		if ml.lastChunk != nil {
			ml.lastChunk = nil
			return ml.Locate(snippet, z, threshold)
		}
		return 0, 0, 0, false
	}

	ml.lastChunk = bestChunk
	return bestX, bestY, float64(bestScore), true
}

func (ml *MinimapLocator) getTileMat(ck navigation.ChunkKey) gocv.Mat {
	if mat, ok := ml.matCache[ck]; ok {
		return mat
	}
	tile, ok := ml.atlas.GetColorTile(ck.X, ck.Y, ck.Z)
	if !ok {
		return gocv.NewMat()
	}
	mat := rgbaToMat(tile)
	ml.matCache[ck] = mat
	return mat
}

// rgbaToMat converts an image.RGBA to a GoCV Mat (RGB, 3-channel).
func rgbaToMat(img *image.RGBA) gocv.Mat {
	bounds := img.Bounds()
	rows := bounds.Dy()
	cols := bounds.Dx()
	mat := gocv.NewMatWithSize(rows, cols, gocv.MatTypeCV8UC3)
	for y := 0; y < rows; y++ {
		for x := 0; x < cols; x++ {
			off := y*img.Stride + x*4
			r := img.Pix[off+0]
			g := img.Pix[off+1]
			b := img.Pix[off+2]
			mat.SetUCharAt(y, x*3+0, r)
			mat.SetUCharAt(y, x*3+1, g)
			mat.SetUCharAt(y, x*3+2, b)
		}
	}
	return mat
}

func (ml *MinimapLocator) getNeighborChunks(center navigation.ChunkKey) []navigation.ChunkKey {
	var neighbors []navigation.ChunkKey
	cs := navigation.ChunkSize
	for dx := -cs; dx <= cs; dx += cs {
		for dy := -cs; dy <= cs; dy += cs {
			key := navigation.ChunkKey{X: center.X + dx, Y: center.Y + dy, Z: center.Z}
			if _, ok := ml.atlas.ColorChunks[key]; ok {
				neighbors = append(neighbors, key)
			}
		}
	}
	return neighbors
}

